Implemented the "confirm visit" in visit service

anuragSharma1112 opened the pull request:
# Description
Implemented the "confirm visit" feature to validate OTP, update visit approval status, and ensure secure and verified check-ins for visitors.

Fixes # (issue)
NA

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Database change

## How Has This Been Tested?
- [X] Unit Tests
- [ ] Integration Tests
- [ ] Manual Testing (please describe)

## Checklist:
- [ ] I have created ADR (Architecture Decision Record) for this change, in case of architecture or tech stack changes.
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes in other branches have been merged
- [ ] I have updated the relevant application properties files (if needed)
- [ ] I have updated any necessary database migrations (if needed)

## Screenshots (for UX changes):
NA

## Additional Notes:
NA

anuragSharma1112 committed:
Implemented the "confirm visit" in visit service

anuragSharma1112 commented:
Hi @axymthr sir please review and merge pull request.

axymthr commented on web-backend/src/main/java/com/statusneo/vms/service/VisitService.java:
```diff
@@ -103,4 +103,20 @@ public Visit registerVisit(Visitor visitor) {
+        boolean isValid = otpService.validateOtp(visit, otpCode);
+
+        if (isValid) {
+            otpService.markVisitAsVerified(visit);
```
Depending on what `otpService.markVisitAsVerified` does, I think it shouldn't exist. OTP Service shouldn't mark visit as verified, that needs to happen here in VisitService. OTP Service should mark OTP as verified (if we have that DB column), but Visit Service doesn't need to know about, it's part of the `validateOtp` call.
The other action we're missing is updating the _visitor email is now verified_ field, which should happen here in VisitService.

axymthr commented on web-backend/src/main/java/com/statusneo/vms/service/VisitService.java:
```diff
@@ -103,4 +103,20 @@ public Visit registerVisit(Visitor visitor) {
+            visitRepository.save(visit);
+            return true;
+        } else {
+            return false;
```
Here's a flow we need to consider. Somehow this method needs to communicate to calling code whether all verification attempts failed, or we need to retry verification. Now that you're also diving into the Controller code, think about how this flow should work.
Let me know your thoughts. Is there a design flaw here, and we should approach it differently, or does our design support this?

axymthr replied:
@anuragSharma1112 I think I have it, let me know your thoughts.
You have 2 options:
- Throw a checked exception (with metadata) in OTP service and catch it in the Controller and do further processing, OR
- Return a VerificationResult object. This object should contain:
1. bool success
2. bool reattempt
3. String message (for failure or retry)
This should be simple enough yet still flexible. Check bool success in VisitService and return as is if false, else save into DB
What do you think?

anuragSharma1112 commented:
Hi @axymthr sir i think option 2 good to go, and i am on it.

anuragSharma1112 committed:
Implemented the "confirm visit" in visit service

anuragSharma1112 commented:
Hi @axymthr sir,

I've removed the markVisitAsVerified call from OtpService as per earlier feedback to ensure OTPService focuses strictly on OTP handling logic.

The visit approval (visit.setIsApproved(true)) is now solely handled inside VisitService where it belongs.

OtpService.validateOtp(...) now only returns a boolean — VisitService decides what to do with that outcome.

We return a VerificationResult object containing success status, reattempt info, and message. This aligns with your suggestion to keep control flow clean and extensible.

All state updates are encapsulated in VisitService, keeping OTPService stateless and isolated from business concerns.

This design improves separation of concerns, makes the codebase more testable, and follows SRP.

axymthr commented on web-backend/src/main/java/com/statusneo/vms/dto/VerificationResult.java:
Can we convert this to a record? It's just an immutable data holder.

axymthr commented on web-backend/src/main/java/com/statusneo/vms/service/VisitService.java:
```diff
@@ -103,4 +104,23 @@ public Visit registerVisit(Visitor visitor) {
+        visit.setIsApproved(true);
+        visitRepository.save(visit);
+
+        return new VerificationResult(true, false, "OTP verified successfully");
```
Good progress, but don't do this here. VerificationResult should be returned from OTPService.validateOtp(). We want to encapsulate all the logic of re-attempts there. So hasExceededOtpAttempts() doesn't need to be exposed outside, it's all internal. If we have to change no. of re-attempts, then it's just 1 config change inside OTPService and nothing changes here in VisitService.
Simply check the VerificationResult.success flag here.

anuragSharma1112 committed:
Make the verification result record.

anuragSharma1112 commented:
Hi @axymthr sir, 
I converted VerificationResult to a record as it's just an immutable data holder.
Also, I moved OTP validation and retry logic fully into OtpService to encapsulate the behavior and keep the controller/service clean and maintainable.
Please review and merge pull request.

axymthr commented on web-backend/src/main/java/com/statusneo/vms/controller/VisitorController.java:
```diff
@@ -108,9 +109,9 @@ public ResponseEntity<String> validateOtp(
         }
 
-        boolean isValid = otpService.validateOtp(visit, otp);
+        VerificationResult result = otpService.validateOtp(visit, otp);
```
Are we calling validateOTP form here or from visitService? Only one of them should be kept.

axymthr commented on web-backend/src/main/java/com/statusneo/vms/controller/VisitorController.java:
```diff
@@ -108,9 +109,9 @@ public ResponseEntity<String> validateOtp(
-        if (isValid) {
+        if (result.success()) {
             otpService.markVisitAsVerified(visit);
             return ResponseEntity.ok("<p class=\"text-green-600 font-bold\">OTP Verified Successfully!</p>");
```
While you're changing and touching the UI, also move this HTML to a JTE fragment.

anuragSharma1112 committed:
Merge branch 'develop' of https://github.com/StatusNeo/vms into VisitService-for-visitor-confirmation-task-53

# Conflicts:
#	web-backend/src/main/java/com/statusneo/vms/controller/VisitorController.java

anuragSharma1112 committed:
remove validate otp from visitor controller

anuragSharma1112 commented:
Hi @axymthr I have some changes in visit Confirmation please review.

axymthr commented on web-backend/src/main/jte/visitConfirmationResult.jte:
```diff
@@ -0,0 +1,24 @@
+#else
+<p style="color: red;">Failed to confirm visit: ${result.message()}</p>
+#if(result.reattempt())
+<p>You can try again.</p>
```
Don't hard code the message here, it defeats the point of sending a verification result message. 
If reattempt is allowed, the backend message should contain that.
If it's not allowed, it would contain that.
The issue for me is, how are they going to try again?
We need to provide a 'Resend OTP' option somehow.
Or should we auto-resent OTP in case of failure?
Think through the behavior before building the UI flow.

anuragSharma1112 committed:
fix resend otp logic

anuragSharma1112 commented:
Hi @axymthr sir, i have fixed the resend otp issue and clean up the code, please review and merge.

axymthr commented:
@anuragSharma1112 please rebase your branch from develop. Looks like there are merge conflicts.
